kadmin/offset_lister.go

```diff
@@ -51,18 +63,53 @@ func (ka *SaramaKafkaAdmin) doListOffsets(group string, offsetsChan chan []Topic
 	listResult, err := ka.admin.ListConsumerGroupOffsets(group, nil)
 	if err != nil {
 		errChan <- err
+		return
+	}
+
+	totalPartitions := 0
+	for _, m := range listResult.Blocks {
+		totalPartitions += len(m)
 	}
 
-	var topicPartitionOffsets []TopicPartitionOffset
+	topicPartitionOffsets := make([]TopicPartitionOffset, 0, totalPartitions)
+
+	resultChan := make(chan TopicPartitionOffset, totalPartitions)
+
+	var wg sync.WaitGroup
 	for t, m := range listResult.Blocks {
 		for p, block := range m {
-			topicPartitionOffsets = append(topicPartitionOffsets, TopicPartitionOffset{
-				Topic:     t,
-				Partition: p,
-				Offset:    block.Offset,
+			wg.Go(func() {
+				func(topic string, partition int32, consumerOffset int64) {
+
+					hwm, err := ka.client.GetOffset(topic, partition, sarama.OffsetNewest)
+					var lag int64
+					if err != nil {
+						hwm = ErrorValue
+						lag = ErrorValue
+					} else {
+						lag = hwm - consumerOffset
+					}
+
+					resultChan <- TopicPartitionOffset{
+						Topic:         topic,
+						Partition:     partition,
+						Offset:        consumerOffset,
+						HighWaterMark: hwm,
+						Lag:           lag,
+					}
+				}(t, p, block.Offset)
```

Comment:
This code has issues with variable capturing, more info: https://medium.com/@mvinicus/understanding-goroutines-in-go-avoiding-variable-capture-issues-in-loops-1f17dda9bdf3